shop/src/utils/useIsMounted.js

```diff
@@ -0,0 +1,9 @@
+import { useEffect, useRef } from 'react'
+
+export default function useIsMounted() {
+  const isMounted = useRef(true)
+
+  useEffect(() => () => isMounted.current = false, [])
```

Comment:
Hi @baaae,

I don't yet have a firm grasp on this code pattern. Could you please help me understand it by describing it via code comments, for example?

Reply:
Sure. There's two thing going on here. 

One, we need to know when the component has unmounted. We do that by setting a cleanup function in the `useEffect`. We make sure the cleanup function fires only once, on unmount, by setting an empty dependency array.

Two, we need a way to pass around the value indicating wether the component is mounted or unmounted. With class components we could assign that value to an instance variable and use it throughout the lifecycle of the component. We're not using class components so even though that's a more elegant solution we can't use it here. Instead we use a [ref and treat it like an instance variable](https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables). 

Overall, when the component first mounts it sets the ref `isMounted.current = true` and when it unmounts it sets `isMounted.current = false`.

Hope that clears things up. Let me know if something is still fuzzy